R/tags.R

```diff
@@ -928,6 +988,18 @@ as.tags.shiny.tag.list <- function(x, ...) {
   x
 }
 
+#' @export
+as.tags.shiny.tag.function <- function(x, ...) {
+  y <- x()
+  # as.tags() doesn't currently have a method for a list of dependencies
+  if (is.list(y) && all(vapply(y, is_html_dependency, logical(1)))) {
+    # In order for S3 dispatch to work properly, this can't simply be `lapply(y, as.tags)`
+    lapply(y, function(z) as.tags(z))
+  } else {
+    as.tags(y)
+  }
+}
```

Comment:
While this isn't technically equivalent, I don't see the reason for _not_ calling `as.tags` for each element.
```suggestion
#' @export
as.tags.list <- function(x, ...) {
  tagList(lapply(x, function(z) {
    # only recurse if it contains an html_dependency
    if (is_html_dependency(z)) {
      return(as.tags(z))
    }
    # maintain backwards compatibility by NOT calling `as.tags(z)
    z
  }))
}
#' @export
as.tags.shiny.tag.function <- function(x, ...) {
  # evaluate and convert to tags
  as.tags(x())
}
```

Reply:
I like the idea, but I'm worried about what this means for the `if (is.list(x) && !isTagList(x))` case...

Reply:
There already is a method called `as.tags.shiny.tag.list(x)` which is the class that `isTagList(x)` checks for.  

So I believe this case is covered already and that historical code path was a dead code path.

Reply:
Ok, yea, but that code path isn't dead. In fact, if you change `unclass(x)` to `x` then you get:

```r
> tagify(structure(list(), class = "foo"))
Error: evaluation nested too deeply: infinite recursion / options(expressions=)?
```

So, _I think_ that line is there simply as a way out of the crazy recursion logic in `tagify()`...it seems like it'd be better to avoid the recursion in that case by just wrapping the list in a `tagList()` like you've suggested here.

BTW, it also seems like we can get away without doing the `lapply()` you've suggested here, thanks to the recursive nature of `tagify()`

Reply:
Handled in 03d79ec and 740c3a3